arcgis-ios-sdk-samples/Geometry/List transformations/ListTransformationsViewController.swift

```diff
@@ -0,0 +1,178 @@
+            let outputSR = map.spatialReference else { return }
+        
+        // if orderByMapExtent is on, use the map extent when retrieving the transformations
+        if orderByMapExtent.isOn {
+            datatumTransformations = AGSTransformationCatalog.transformationsBySuitability(withInputSpatialReference: inputSR, outputSpatialReference: outputSR, areaOfInterest: mapView.visibleArea?.extent)
+        }
+        else {
+            datatumTransformations = AGSTransformationCatalog.transformationsBySuitability(withInputSpatialReference: inputSR, outputSpatialReference: outputSR)
+        }
+        
+        defaultTransformation = AGSTransformationCatalog.transformation(forInputSpatialReference: inputSR, outputSpatialReference: outputSR)
+        tableView.reloadData()
+    }
+    
+    func setPEDataURL() {
+        // find the PEDataRuntime folder from the documents directory
+        // added using iTunes
+        let path = NSSearchPathForDirectoriesInDomains(FileManager.SearchPathDirectory.documentDirectory, FileManager.SearchPathDomainMask.userDomainMask, true)
+        let subpaths = FileManager.default.subpaths(atPath: path[0])!
+        
+        // search for PEDataRuntime matches
+        let predicate = NSPredicate(format: "SELF MATCHES %@", "PEDataRuntime")
+        let peDataRuntimePaths = subpaths.filter({ (objc) -> Bool in
+            return predicate.evaluate(with: objc)
+        })
+
+        // use first matching path as path to PE data
+        if let documentPEDataRuntime = peDataRuntimePaths.first {
+            // found "PEDataRuntime" folder, create full url
+            let peDataURL = URL(fileURLWithPath: path[0]).appendingPathComponent(documentPEDataRuntime)
+            try? AGSTransformationCatalog.setProjectionEngineDirectory(peDataURL)
+        }
+        
+        setupTransformsList()
+    }
+
+    @IBAction func oderByMapExtentValueChanged(_ sender: Any) {
+        setupTransformsList()
+    }
+
+    //MARK: - TableView data source
+
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return datatumTransformations.count
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: "DatumTransformCell", for: indexPath)
+
+        // get the selected transformation
+        let transformation = datatumTransformations[indexPath.row]
+
+        // if we're missing the grid files, detail which ones
+        var files = ""
+        if transformation.isMissingProjectionEngineFiles {
+            files = "Missing grid files"
+
+            if let gt = transformation as? AGSGeographicTransformation {
+                gt.steps.forEach { (step) in
+                    if step.isMissingProjectionEngineFiles {
+                        files.append(": " + step.projectionEngineFilenames.joined(separator: ","))
```

Comment:
If there are multiple steps missing files, the message would be something like this (with too many colons):

> Missing grid files: a.gsb: b.gsb

You could use `flatMap` to get an array of all the missing files and then `join` them:

```swift
cell.textLabel?.text = {
    if transformation.isMissingProjectionEngineFiles,
        let geographicTransformation = transformation as? AGSGeographicTransformation {
        let files = geographicTransformation.steps.flatMap { (step) -> [String] in
            step.isMissingProjectionEngineFiles ? step.projectionEngineFilenames : []
        }
        return "Missing grid files: \(files.joined(separator: ", "))"
    } else {
        return ""
    }
}()
```

Reply:
Nice.  Done.